### PR TITLE
Added new features to script window

### LINKED ISSUE #6541: Using rtf editor for the R-Instat script window
Changing to an rtf editor for the script window is related to the need for bold/colour, etc in the files in the script window.  That need currently relates to an issue where dialogues can choose to emphasise a word/command in the file.

However, it could be that the choice of editor could help in other tasks too.  I mwention 2 points here that might be considered when choosing the editor.

1) We do want R-Instat to become multi-platform.  I assume we can choose an editor that will be ok on each system.
2) We want to be able to edit (awkward) html (xml) files.  That's to be able to locate data (hidden) within html files.  
3) We (I) don't use the log window much.  And that is read-only.  But presumably we may (or may not?) use the same (now) rtf viewer for the log and script window.

I am assuming slotting in the appropriate rtf viewer might be quite easy.  I don't know if the choice of rtf viewer is easy or not.  Our philosophy is "open by default".  So we will use a free (and open source) editor if that is available.  But we would not object to buying an editor if there were a good reason to do so - as long as it could then be included free.
 
I wonder about notepad++?  As of 2020 it runs under wine, which we plan to use for our multi-platform work.  Maybe @ChrisMarsh82 can comment?

And, I don't know if it is related much, but I would also also like a viewer - and ideally an editor for pdf files.  I would like a viewer, partly because information about data sets in our library will often be in pdf files.  We should be able to see them?  Then, an editor would be useful to be able to get tables of data from within a pdf file.

This links (I assume) with our new facility to paste into a new data frame.  Might that be possible - perhaps (like Excel) via a wizard?

### LINKED ISSUE #6645: Add line numbers in the script window
I was in a situation the other day when using the script window in R-Instat with someone else over zoom. We were discussing code, and it was difficult to clarify the line of code we were each referring to.

As a solution, can line numbers be added into the script window?

I noticed it was discussed in #4381, alongside other improvements. The other suggestions are either implemented or in a separate issue now, but I could not find a place where line numbers were discussed again.

### PR DESCRIPTION
**Pull request by lloyddewit**
This PR is draft, not yet ready to review.

Fixes #6645
Fixes #6541
Fixes (partly) #4932 

Fixes (potentially):

- #7632
- #7435
- #7014
- #6744
- #6720
- #6683
- #6619
- #6178
- #5807
- #4444
- #4381
- #4216
- #3567
- #3566

### CODE REVIEW

**Comment by lloyddewit** [comment]
@rdstern @Patowhiz 
This PR uses the 'ScintillaNET' package. This package provides more advanced editor features than the Windows Forms 'TextBox' control. This allows us to add features such as syntax highlighting and line numbers. However, I don't know how well 'ScintillaNET' would run in the R-Instat Linux version.
Who would be the best person to test this PR branch in Linux?

**Comment by lloyddewit** [comment]
@rdstern 
I added some of the key editor features requested by @lilyclements :
- Fixed width font
- Syntax highlighting
- Ensured that typical editor functions like cut/copy/paste/undo/redo/select etc. all work in a standard way with standard keybopard shortcuts
- Improved right mouse click menu

If you agree, then I think this would be a good time to merge. I would implement the next improvements in a new PR.
Please could you test?

Possible next improvements:
- make 'ctrl-enter' execute the R statement at the cursor (rather than just the line), this should fix the current bugs with multi-line statements 
- add tabbed windows
- add auto close brackets, close quotes and indentation
- add reformat option to tidy indentation and line breaks (including splitting long lines)
- tidy up file open/save/save-as functionality
- add auto-complete
- find/replace dialog
- display cursor row/column at bottom of editor
- ensure commands on R-Instat 'Edit' menu work as expected in script window
- ensure controls translate to French and other languages
- update help files
- test on Unix

Do you have a preference regarding which I should implement first (I can provide time estimates if needed)?

If you want to compare against RStudio, then the RStudio editor functionality is described below.

https://support.rstudio.com/hc/en-us/articles/200484448-Editing-and-Executing-Code-in-the-RStudio-IDE
https://support.rstudio.com/hc/en-us/articles/200710523-Navigating-Code-in-the-RStudio-IDE
https://support.rstudio.com/hc/en-us/articles/205273297-Code-Completion-in-the-RStudio-IDE
https://support.rstudio.com/hc/en-us/articles/200484568-Code-Folding-and-Sections-in-the-RStudio-IDE

**Review by rdstern — APPROVED** [review]
@lloyddewit this is looking very nice indeed.  I have searched for something to complain about, but it seems to work fine!

**Comment by rdstern** [comment]
@lloyddewit I don't have strong views about the order of the items in the new pull request.  So very happy for you to decide on the most efficient logical order.

**Comment by lloyddewit** [comment]
This is just a note for myself that I could test the script window with code from issue #7889.

File: instat/ucrScript.vb [hunk 1]
```diff
@@ -144,134 +164,133 @@ Public Class ucrScript
...
+                       "The file may be in use by another program or you may not have access to write to the specified location.",
+                       vbExclamation, "Load Script")
+            End Try
         End Using
     End Sub
 
     Private Sub mnuCopy_Click(sender As Object, e As EventArgs) Handles mnuCopy.Click
-        If txtScript.SelectionLength > 0 Then
+        If txtScript.SelectedText.Length > 0 Then
             CopyText()
```
Comment (N-thony) [inline instat/ucrScript.vb:204]: txtScript.Copy()? to make it consistent with line 211 and get rid of CopyText()?

**Review by N-thony — APPROVED** [review]

File: instat/ucrScript.vb (hunk 1 above)
Comment (lloyddewit) [reply instat/ucrScript.vb:204]: Thanks, I understand your point.
`Copytext()` is also called by `frmMain`. Therefore, we cannot delete it.
If we don't call `CopyText()` on line 204, then we also need to call `EnableDisableButtons()` which adds an extra line of duplicated code.
So on balance, I'll leave the above code unchanged. I hope that's OK.

**Pull request merged**